### DISCUSSION TITLE
[Proposal] Deconstruct in place - a tiny deconstruction enhancement

### DISCUSSION THREAD

**Discussion by iam3yal** [dotnet/csharplang#7200] [Language Ideas] [answered]
Say I have a type `T` with the properties P1 and P2 that implements a `Deconstruct` method and I now want to deconstruct the instance I have to do something like this:
```csharp
var t = new T()
var (p1, p2) = t;
```
I propose the following syntax where the deconstruction follows the instantiation of the object so the instance of the type is always first in order so with this proposal we could merge it all into a single line like this:
```csharp
var (t, p1, p2) = new T()
```
**The semantics are equivalent exactly to the above.**

In my own codebase I have a Playwright wrapper that atm I have to do something like this:
```csharp
await using var playwright = await PlaywrightFactory.CreateAsync();

var (_, context) = playwright;
```
And I wish I could express this as a single line.
```csharp
await using var (playwright, _, context) = await PlaywrightFactory.CreateAsync();
```
The obvious alternative is accessing the property directly but then you have to repeat yourself so in the Playwright example I have to go through the `playwright` instance each time I want to access the `Context` property or make my code a bit more verbose than necessary (in my opinion) and use deconstruction to access the `context` directly so between these two options my proposal stands.

**Comment by huoyaoyuan** [comment]
The syntax is very ambiguous comparing to normal deconstruction with more elements.
In fact, you can already write an extension `Deconstruct` method to achieve that:

https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEAsBuWsQGZkQDZkxEBhRAb1kRuTxSJFUQFkAKAaQFMBPANQCGAGwCuXAAoCAlgCcAPFIB2AFwA0+eAD5EAawBuABwCU1WlRi1LiPQJmI2AexHKOh9TvV6jiALy7DmBa0AL6mNGF0+IzMACJcAMYOigDOyjIi8cpyHOp8mmzKABZSyYjc/MJikrLZudr6BqoRVi2tbe00TsplvIKiEtLyOYh5iF2ujc0d0zO0XWW6vE1Bs6sz83zWlVwmK5RTnc4Tvv4GgW1sOktb/d5+DechsMFAA==

**Reply by iam3yal** [reply]
No you can't write an extension to achieve this because the syntax is illegal today and the syntax is not very ambiguous, in fact, it isn't ambiguous at all and your example is different than what I'm proposing because you're deconstructing an existing instance of `KeyValuePair<K, V>` whereas what I'm proposing takes place only when a `new`  object is created or at instantiation, not to mention that async methods can't have `ref`, `in` or `out` parameters today.

**Comment by DavidArno** [comment]
```cs
var (t, (p1, p2)) = new T() is var t_ ? (t_, t_) : (t_, t_);
```
😈

**Reply by btnlq** [reply]
```cs
var (p1, p2) = new T() is var t ? t : t;
```

**Reply by DavidArno** [reply]
That's far too simple, @btnlq 😆

**Reply by iam3yal** [reply]
We should definitely embrace this form and write all of our code with such cleverness. 😆😉

You should try doing the same with an `async` `Deconstruct` method or a `using` declaration, the point is to allow the syntax of instantiation and then deconstruct it in place, I think that yes while it's extremely small quality of life improvement it has merits in my opinion.

**Comment by En3Tho** [comment]
I think I would rather see 
```csharp
var t and (x1, x2) = new T1();
```

and same with Playwright:
```csharp
await using var playwright and (_, context) = await PlaywrightFactory.CreateAsync();
```

This way it's more similar to usual pattern matching. 

F# can already do this in "let" bindings (readonly "var") like 
```fsharp
let tuple & (fst, snd) = (1, 2)
```
But not with "use" tho

P.S. I miss the same feature in TypeScript too.

**Reply by iam3yal** [reply]
@En3Tho I wouldn't mind having this syntax or any other that is more pleasant.

**Comment by HaloFour** [comment]
You can use extension deconstructors to achieve this kind of behavior:

```cs
public static class MyExtensions {
    public static void Deconstruct(this T input, out T output, out int p1, out int p2) {
        output = input;
        (p1, p2) = input;
    }
}
```

Then you can use it like this:

```cs
var (instance, p1, p2) = new T();
```

or even like this:

```cs
// given
public Task<T> SomeAsyncMethod() { ... }

// then
var (instance, p1, p2) = await SomeAsyncMethod();
```

Although that does require writing an extension per type you want to deconstruct in this manner.

**Reply by iam3yal** [reply]
@HaloFour You can't do that because of two issues 1) You can't use a `using` declaration and 2) You can't have an `async` `Deconstruct` method due to its `out` parameters.

**Reply by HaloFour** [reply, marked as answer]
Correct, you cannot combine `using` with deconstruction.

You don't need an async `Deconstruct`, you only need to `Deconstruct` the awaited result.  The above code shows how you can deconstruct the awaited result of an async method and that works fine.  But you can't combine it with `using`.

IMO, I don't think a language feature that specifically supports packing `await`, `async using` and deconstruction into a single line is a good idea.  I imagine that the scenario in which it would come up is incredibly niche and shouldn't be encouraged.  I also don't like silently supporting deconstruction that echos the rvalue.  If that can be composed out of other building blocks, though, I could be on board.  e.g., being able to use a `using` directive on a deconstructed value:

```cs
var (using instance, p1, p2) = await SomeAsyncMethod();
```

**Reply by iam3yal** [reply]
@HaloFour Gotcha, thank you for elaborating.
